src/rt/cpp/when.h

```diff
@@ -124,6 +124,35 @@ namespace verona::cpp
       }
     }
 
+    AccessBatch(cown_ptr_span<T>&& ptr_span) : is_move(true)
+    {
+      // Allocate the actual_cown and the acquired_cown array
+      // The acquired_cown array is after the actual_cown one
+      size_t actual_size =
+        ptr_span.length * sizeof(ActualCown<std::remove_const_t<T>>*);
+      size_t acq_size =
+        ptr_span.length * sizeof(acquired_cown<std::remove_const_t<T>>);
+      span.array = reinterpret_cast<ActualCown<std::remove_const_t<T>>**>(
+        snmalloc::ThreadAlloc::get().alloc(actual_size + acq_size));
+
+      for (size_t i = 0; i < ptr_span.length; i++)
+      {
+        span.array[i] = ptr_span.array[i].allocated_cown;
+      }
+      span.length = ptr_span.length;
+
+      acq_array =
+        reinterpret_cast<acquired_cown<T>*>((char*)(span.array) + actual_size);
+
+      for (size_t i = 0; i < ptr_span.length; i++)
+      {
+        new (&acq_array[i]) acquired_cown<T>(*ptr_span.array[i].allocated_cown);
+      }

```

Comment:
Could you factor out the code duplication here from the above constructor. Thanks